torch/csrc/jit/codegen/cuda/lower_predicate_elimination.cpp

```diff
@@ -241,6 +263,99 @@ class PredicateChcker : public IterVisitor {
     return false;
   }
 
+  bool predicateSharedMemAccess(Expr* expr) const {
+    // This is initial step to gradually remove predicates around
+    //  sharedmem access in suitable situations.
+    // Using an additional variable to track the predicate-on reasons
+    //  when the predicate around shared mem cannot be removed.
+    for (auto consumer : ir_utils::filterByType<TensorView>(expr->outputs())) {
+      for (auto producer : ir_utils::filterByType<TensorView>(expr->inputs())) {
+        if (producer->getMemoryType() == MemoryType::Shared ||
+            consumer->getMemoryType() == MemoryType::Shared) {
+          if (needSharedMemPredicate(producer, consumer)) {
+            return true;
+          }
+        }
+      }
+    }
+    return false;
+  }
+
+  // Check for conditions where the predicate cannot be removed
+  //  when either producer or consumer is in shared memory.
+  bool needSharedMemPredicate(TensorView* producer, TensorView* consumer)
+      const {
+    // Indexing is based on consumer leaf ids so check the consumer.
+
+    // If consumer schedule contains in-exact thread parallel
+    //  dimensions, need to predicate against out of bound
+    //  shared memory access by out of bound threads.
+    if (!isExactParallelSharedMemAccess(consumer)) {
+      return true;
+    }
+
+    for (auto id : consumer->domain()->domain()) {
+      // TODO: (Enable in a follow up)
+      //  smem predicate removal with init would break unroll and unswitch,
+      //  eg. as in issue 1133, so disabling this removal pattern for now.
+      if (id->getParallelType() == ParallelType::Unroll ||
+          id->getParallelType() == ParallelType::Unswitch) {
+        return true;
+      }
+
+      // TODO: （Enable in a follow up)
+      //  This cannot yet be removed since smem initialization needs to be
+      //  handled specially, e.g. as in smem_reduce test. Will be able to
+      //  lift this one once the generic pred removal pass with fusion
+      //  traversal is ready.
+      auto consumer_def = consumer->definition();
+      if (ir_utils::isReductionOp(consumer_def)) {
+        if (producer->getMemoryType() == MemoryType::Shared) {
+          return true;
+        }
+      }
+    }
+
+    return false;
+  }
+
+  // Utility to find the leaf iterdomains of the given
+  //   tensor view that will be treated as "zero loops"
+  //   in the indexing pass.
```

Comment:
Can you please mention here that `indexMapFromTV` should be also looked at about zero loops?